### DISCUSSION TITLE
Intermittent Connection Pool Timeouts with Managed PostgreSQL and PgBouncer

### DISCUSSION THREAD

**Discussion by cureau** [vercel/community#4756] [Help] [answered]
### Summary

We are experiencing intermittent timeouts when trying to fetch new connections from the connection pool on my PostgreSQL database. The error message I receive is:

`Timed out fetching a new connection from the connection pool. More info: http://pris.ly/d/connection-pool (Current connection pool timeout: 10, connection limit: 5)`

Given that Vercel's documentation states that PgBouncer allows for up to 10,000 simultaneous connections and up to 450 direct connections on pro, the connection limit of 5 seems to be a bottleneck. The suggested fix is to set `connect_timeout=0` and increase the `connection_limit` parameter.

However, Vercel Postgres manages our environment variables for us. Nor can we append params to Prisma's connection configuration.

For context, we are using Prisma ORM with a Next.js application and the Vercel serverless postgres. This is affecting production users so any immediate help is appreciated.

Thank you!

### Steps to Reproduce

1. Connect prisma to vercel serverless postgres

```
datasource db {
  provider   = "postgresql"
  url        = env("POSTGRES_PRISMA_URL") // uses connection pooling
  directUrl  = env("POSTGRES_URL_NON_POOLING") // uses a direct connection
  extensions = [vector]
}
```

2. Open more than 5 connections at once

**Comment by cureau** [comment]
There are two potential issues in the above:

1. Pooling isn't working properly. This could be with the pgbouncer config on neon/vercel's side. Or it could be with prisma. None of our code used postgres features that were unsupported by postgres transaction mode.

2. Regardless of the handling, there's ambiguity on how to handle our prisma config via vercel's managed postgres. For instance, in the  [Vercel documentation here](https://vercel.com/docs/storage/vercel-postgres/vercel-postgres-error-codes#p2024:-timed-out-fetching-a-new-connection-from-the-connection-pool) the suggestion is to increase `connection_limit`

However, since Vercel's `POSTGRES_PRISMA_URL` environmental var is handled by the managed service, there's no way to modify this directly. Instead, we might append `?connection_limit` and other paras to the prisma schema setup (i.e., `datasource db`) but (a) it's "hacky" and (b) it could break with future changes to vercel's connection string (presuming that method works at all).

```
datasource db {
  provider   = "postgresql"
  url        = env("POSTGRES_PRISMA_URL") // uses connection pooling
  directUrl  = env("POSTGRES_URL_NON_POOLING") // uses a direct connection
}
```

In short:

1. I'm not sure why pooling isn't always handle correctly
2. In any case, how does vercel suggest we manage `connection_limit`, timeout, etc given the managed nature of connection strings?

**Reply by isaldns** [reply]
this is exactly my problem now

**Reply by austinkelleher** [reply]
We are experiencing the same issue. @cureau Thank you for capturing the details. Did you ever work out a solution?

**Reply by timothywadecook** [reply]
I am experiencing the same issue. Any updates?

**Reply by Schmell** [reply]
adding `?connect_timeout=0` to prisma schema does not work.
```{.prisma .no-copy}
// DO NOT COPY THIS!!!!
datasource db {
  provider   = "postgresql"
  url        = env("POSTGRES_PRISMA_URL?connect_timeout=0") // uses connection pooling
  directUrl  = env("POSTGRES_URL_NON_POOLING") // uses a direct connection
}
```
This will function on localhost dev , but you will get a "This Serverless Function has crashed." page on Vercel.
_(FYI: you cannot provide string templates or any javascript as an argument to `url`)_

Apparently you don't have to use the Vercel managed connection.
My fix is to create your own .env variable. 
```env
MY_POSTGRES_PRISMA_URL="postgres://default:....?pgbouncer=true&connect_timeout=0&connection_limit=10"
```
add to localhost .env and vercel .env variables.

then change schema.prisma to
```prisma
datasource db {
  provider  = "postgresql"
  url       = env("MY_POSTGRES_PRISMA_URL") // with added connection query params
  directUrl = env("POSTGRES_URL_NON_POOLING") // no change as it is direct anyway
}
```

### NOTE:
In my personal experience is that `connect_timeout=0` does not solve the problem, but with this method you can just increase the number until it meets your requirements. Also there are other query params you can use that might be more relevant to your requirements. (_eg: `pool_timeout` etc.._)

**Reply by runningjustin** [reply]
I am having the same problem. Thinking of finding a new Postgres service as this is becoming a blocker for me.

Note @Schmell that your appending of connect_timeout to the env variable itself will never work. You'll need to add that parameter outside the env() block. But even then, not sure if appending will work correctly here.

**Reply by cureau** [reply, marked as answer]
The Vercel team eventually responded saying you should manage/overwrite your own connection string. It feels a bit hacky to me and we still get these issues often in production, so I think Vercel could do better here, but so far that's where we're at. Here's the relevant reply from them:

> Vercel Postgres environment variables are static once created. We do not change them over time and it is very unlikely that we’d retroactively update them. The recommendation here is to manage the connection string yourself or use something like `dotenv-expand` to extend the `POSTGRES_PRISMA_URL` environment variable in your application code.

**Reply by maestrodrew** [reply]
In my case, I noticed these errors occurring only after modifying the inactivity timeout period in Vercel Postgres settings. I had modified the value to 60 seconds and started experiencing this issue. After setting it back to the default of 300 seconds, no issues.
